lib/aerospike/client.rb

```diff
@@ -315,32 +315,44 @@ def get_header(key, options = nil)
     #  Read multiple record headers and bins for specified keys in one batch call.
     #  The returned records are in positional order with the original key array order.
     #  If a key is not found, the positional record will be nil.
-    #  The policy can be used to specify timeouts.
-    def batch_get(keys, bin_names = nil, options = nil)
-      policy = create_policy(options, Policy)
+    #  The policy can be used to specify timeouts and protocol type.
 
-      # wait until all migrations are finished
-      # TODO: implement
-      # @cluster.WaitUntillMigrationIsFinished(policy.timeout)
 
-      # same array can be used without sychronization;
-      # when a key exists, the corresponding index will be set to record
+    def batch_get(keys, bin_names=[], options={})
+      policy = create_policy(options, BatchPolicy)
       records = Array.new(keys.length)
-
-      key_map = BatchItem.generate_map(keys)
-
-      batch_execute(keys) do |node, bns|
-        BatchCommandGet.new(node, bns, policy, key_map, bin_names, records, INFO1_READ)
+      # Use old batch direct protocol where batch reads are handled by direct low-level batch server 
+      # database routines.  The batch direct protocol can be faster when there is a single namespace, 
+      # but there is one important drawback.  The batch direct protocol will not proxy to a different 
+      # server node when the mapped node has migrated a record to another node (resulting in not
+      # found record).  
+
+      # This can happen after a node has been added/removed from the cluster and there is a lag 
+      # between records being migrated and client partition map update (once per second).
+
+      # The new batch index protocol will perform this record proxy when necessary.
+      # Default: false (use new batch index protocol if server supports it)    
+      if policy.use_batch_direct 
+        key_map = BatchItem.generate_map(keys)
+
+        batch_execute(keys) do |node, bns|
+          BatchDirectCommandGet.new(node, bns, policy, key_map, bin_names.uniq, records, INFO1_READ)
+        end
+      else
+        policy = create_policy(options, Policy)
+        records = Array.new(keys.length)        
```

Comment:
Why are you re-initializing the policy and records objects here in this branch? They are already being initialized at the beginning of the function.